src/slurm_plugin/clustermgtd.py

```diff
@@ -964,6 +989,93 @@ def _is_node_in_replacement_valid(self, node, check_node_is_valid):
             return not time_is_expired if check_node_is_valid else time_is_expired
         return False
 
+    def _handle_ice_nodes(self, active_nodes: list, unhealthy_dynamic_nodes: list):
+        """Handle nodes failed with insufficient capacity."""
+        # get insufficient capacity compute resource and nodes mapping
+        ice_compute_resources_and_nodes_map = self._get_unhealthy_ice_nodes(unhealthy_dynamic_nodes)
+        if ice_compute_resources_and_nodes_map:
+            self._update_insufficient_capacity_compute_resources(ice_compute_resources_and_nodes_map)
+            self._reset_timeout_expire_compute_resources(ice_compute_resources_and_nodes_map)
+        self._set_ice_compute_resources_to_down(active_nodes)
+
+    def _update_insufficient_capacity_compute_resources(self, ice_compute_resources_and_nodes_map: dict):
+        """Add compute resource to insufficient_capacity_compute_resources if node is ICE node."""
+        for compute_resource, data in ice_compute_resources_and_nodes_map.items():
+            if compute_resource not in self._insufficient_capacity_compute_resources:
+                self._insufficient_capacity_compute_resources[compute_resource] = ComputeResourceFailureEvent(
+                    self._current_time, data.error_code
+                )
+
+    def _reset_timeout_expire_compute_resources(self, ice_compute_resources_and_nodes_map: dict):
+        """Reset compute resources which insufficient_capacity_timeout expire."""
+        # Find insufficient_capacity_timeout compute resources
+        if not self._insufficient_capacity_compute_resources:
+            return
+        log.info(
+            "The following compute resources are in down state due to insufficient capacity: %s, "
+            "compute resources will be reset after insufficient capacity timeout (%s seconds) expire",
+            self._insufficient_capacity_compute_resources,
+            self._config.insufficient_capacity_timeout,
+        )
+        timeout_expired_compute_resources = self._find_insufficient_capacity_timeout_expire_compute_resources()
+
+        # Reset nodes which insufficient capacity timeout expire
+        if timeout_expired_compute_resources:
+            self._reset_insufficient_capacity_timeout_expire_nodes(
+                timeout_expired_compute_resources, ice_compute_resources_and_nodes_map
+            )
+
+    def _set_ice_compute_resources_to_down(self, active_nodes: list):
+        """Set powered_down nodes which belong to insufficient capacity compute resources to down."""
+        if not self._insufficient_capacity_compute_resources:
+            return
+        nodes_to_down = {}
+        for node in active_nodes:
```

Comment:
here we are looping over all cluster nodes if a single compute resource has ICE. Can we think of a way to group nodes by partition/compute_resource so that we can easily get nodes from the interested compute resources?

Ideally we would like to build the required data structures the first time we loop over all nodes. If I'm not mistaken currently we do it in `_retrieve_scheduler_partitions`. One possibility, but please do investigate alternatives, is to have `_retrieve_scheduler_partitions` return an additional dict where you have CR to node mapping.

Reply:
Make `_retrieve_scheduler_partitions` also return queue_compute_resource_nodes_map.
But retrieving  queue_compute_resource_nodes_map inside the function make it looks like the function is doing two things. Do you think we should rename the function? Do you have a suggestion on a better name on this function? Thank you!
```
    @staticmethod
    def _retrieve_scheduler_partitions(nodes):
        try:
            ignored_nodes = []
            queue_compute_resource_nodes_map = {}
            partitions_name_map = ClusterManager._get_partition_info_with_retry()
            log.debug("Partitions: %s", partitions_name_map)
            for node in nodes:
                if not node.partitions or any(p not in partitions_name_map for p in node.partitions):
                    # ignore nodes not belonging to any partition
                    ignored_nodes.append(node)
                else:
                    for p in node.partitions:
                        partitions_name_map[p].slurm_nodes.append(node)
                    queue_compute_resource_name = node.queue_name + "-" + node.compute_resource_name
                    queue_compute_resource_nodes_map.setdefault(queue_compute_resource_name, []).append(node)
            if ignored_nodes:
                log.warning("Ignoring following nodes because they do not belong to any partition: %s", ignored_nodes)
            return partitions_name_map, queue_compute_resource_nodes_map
        except Exception as e:
            log.error("Failed when getting partition/node states from scheduler with exception %s", e)
            raise
```

Reply:
something like `parse_scheduler_nodes_data`?